src/cross_section/fillet2d.cpp

```diff
@@ -0,0 +1,1311 @@
+  t = -1;
+
+  if (length(p2 - p1) < EPSILON) {
+    t = 0;
+    return length(p1 - p);
+  }
+
+  vec2 d = p2 - p1;
+
+  t = dot(p - p1, d) / dot(d, d);
+
+  vec2 closestPoint;
+
+  if (t < 0) {
+    t = 0;
+    closestPoint = p1;
+  } else if (t > 1) {
+    t = 1;
+    closestPoint = p2;
+  } else {
+    closestPoint = p1 + t * d;
+  }
+
+  return length(closestPoint - p);
+}
+
+// Check if line segment intersect with a circle
+bool intersectCircleSegment(const vec2& p1, const vec2& p2, const vec2& center,
+                            double radius) {
+  if (length(p2 - p1) < EPSILON)
+    return length2(p1 - center) < (radius + EPSILON) * (radius + EPSILON);
+
+  double t;
+  return distancePointSegment(center, p1, p2, t) < (radius + EPSILON);
+};
+
+// Calculate line segment intersections with a circle
+int intersectCircleSegment(const vec2& p1, const vec2& p2, const vec2& center,
+                           double radius, std::array<vec2, 2>& intersections) {
+  vec2 d = p2 - p1;
+  vec2 f = p1 - center;
+
+  double a = dot(d, d);
+  double b = 2 * dot(f, d);
+  double c = dot(f, f) - radius * radius;
+
+  double discriminant = b * b - 4 * a * c;
+
+  if (discriminant < 0) return 0;  // No intersection
+
+  double sqrtDisc = sqrt(discriminant);
+  double t1 = (-b - sqrtDisc) / (2 * a);
+  double t2 = (-b + sqrtDisc) / (2 * a);
+
+  int count = 0;
+  if (t1 >= -EPSILON && t1 <= 1 + EPSILON) {
+    intersections[count] = p1 + d * t1;
+    count++;
+  }
+  if (t2 >= -EPSILON && t2 <= 1 + EPSILON && abs(t2 - t1) > EPSILON) {
+    intersections[count] = p1 + d * t2;
+    count++;
+  }
+
+  return count;
+};
+
+// Check if line segment intersect with another line segment
+bool intersectSegmentSegment(const vec2& p1, const vec2& p2, const vec2& p3,
+                             const vec2& p4, double& t, double& u) {
+  t = -1;
+  u = -1;
+
+  double det = cross(p2 - p1, p4 - p3);
+
+  if (std::abs(det) < EPSILON) {
+    // Parallel
+    return false;
+  }
+
+  double num_t = cross(p3 - p1, p4 - p3);
+  double num_u = cross(p3 - p1, p2 - p1);
+
+  t = num_t / det;
+  u = num_u / det;
+
+  // Check if the intersection point inside line segment
+  if ((t >= 0.0 - EPSILON && t <= 1.0 + EPSILON) &&
+      (u >= 0.0 - EPSILON && u <= 1.0 + EPSILON)) {
+    // Inside
+    return true;
+  } else {
+    // Outside -> No intersection
+    t = u = 0;
+
+    return false;
+  }
+};
+
+int calculatePointPointTangentCircles(const vec2& p1, const vec2& p2,
+                                      double radius,
+                                      std::array<vec2, 2>& circleCenters) {
+  double dx = p2.x - p1.x;
+  double dy = p2.y - p1.y;
+
+  double distance = sqrt(dx * dx + dy * dy);
+
+  vec2 v = p2 - p1;
+
+  // Find midpoint between the two points
+  vec2 midpoint((p1.x + p2.x) / 2.0, (p1.y + p2.y) / 2.0);
+
+  // If distance equals 2*radius, there's exactly one solution (midpoint)
+  if (abs(length(v) - 2 * radius) < 1e-9) {
```

Comment:
For things like `abs`, make sure you are using `std::abs`. The one with no namespace is C's `abs`, which deals with integers...

Reply:
Thanks!